internalshared/listenerutil/custom_response_headers.go

```diff
@@ -0,0 +1,160 @@
+package listenerutil
+
+import (
+	"fmt"
+	"net/http"
+	"net/textproto"
+	"strconv"
+	"strings"
+
+	"github.com/hashicorp/vault/internalshared/configutil"
+)
+
+type ListenerCustomHeaders struct {
+	Address string
+	StatusCodeHeaderMap map[string][]*CustomHeader
+	// ConfiguredHeadersStatusCodeMap field is introduced so that we would not need to loop through
+	// StatusCodeHeaderMap to see if a header exists, the key for this map is the headers names
+	ConfiguredHeadersStatusCodeMap map[string][]string
+}
+
+type CustomHeader struct {
+	Name string
+	Value string
+}
+
+// ChangeListenerAddress is used for tests where the listener address (at least the port)
+// is chosen at random
+func (l *ListenerCustomHeaders) ChangeListenerAddress(la string) {
+	l.Address = la
+	return
+}
+
+func (l *ListenerCustomHeaders) SetCustomResponseHeaders(w http.ResponseWriter, status int) {
```

Comment:
Any suggestion on how to handle errors in this function? should we return them, or somehow log them? In the one hand, I am not sure if it should interrupt request processing, on the other hand, it should not silently fail. So, please provide me with your suggestions on this.